File: 2081-sum-of-k-mirror-numbers/2081-sum-of-k-mirror-numbers.py
class Solution:
    def kMirror(self, k: int, n: int) -> int:
        def is_pal(x: int) -> bool:
            digit = list()
            while x:
                digit.append(x % k)
                x //= k
            return digit == digit[::-1]

        left, cnt, ans = 1, 0, 0
        while cnt < n:
            right = left * 10
            # op = 0 indicates enumerating odd-length palindromes
            # op = 1 indicates enumerating even-length palindromes
            for op in (0, 1):
                # enumerate i
                for i in range(left, right):
                    if cnt == n:
                        break

                    combined = i
                    x = i // 10 if op == 0 else i
                    while x:
                        combined = combined * 10 + x % 10
                        x //= 10
                    
                    if is_pal(combined):
                        cnt += 1
                        ans += combined
            left = right
        
        return ans

File: 2081-sum-of-k-mirror-numbers/2081-sum-of-k-mirror-numbers.cpp
class Solution {
private:
    int digit[100];

public:
    long long kMirror(int k, int n) {
        auto isPalindrome = [&](long long x) -> bool{
            int length = -1;
            while (x){
                digit[++length] = x % k;
                x /= k;
            }

            for (int i = 0, j = length; i < j; ++i, --j){
                if (digit[i] != digit[j]) return false;
            }

            return true;
        };

        int left = 1, cnt = 0;
        long long ans = 0;

        while (cnt < n){
            int right = left * 10;
            // Op = 0 indicates enumerating odd-length palindromes
            // Op = 1 indicates enumerating even-length palindromes
            for (int op = 0; op < 2; ++op){
                for (int i = left; i < right && cnt < n; ++i){
                    long long combined = i;
                    int x = (op == 0)? i / 10 : i;
                    while (x){
                        combined = combined * 10 + x % 10;
                        x /= 10;
                    }
                    if (isPalindrome(combined)){
                        ++cnt;
                        ans += combined;
                    }
                }
            }
            left = right;
        }

        return ans;
    }
};